platform/http_client_curl.cpp

```diff
@@ -0,0 +1,261 @@
+
+#include "base/assert.hpp"
+#include "base/exception.hpp"
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+#include "boost/uuid/uuid_generators.hpp"
+#include "boost/uuid/uuid_io.hpp"
+
+#include "std/array.hpp"
+#include "std/fstream.hpp"
+#include "std/sstream.hpp"
+#include "std/vector.hpp"
+
+#include <cstdio>    // popen, tmpnam
+
+#ifdef _MSC_VER
+#define popen _popen
+#define pclose _pclose
+#else
+#include <unistd.h>  // close
+#endif
+
+namespace
+{
+DECLARE_EXCEPTION(PipeCallError, RootException);
+
+struct ScopedRemoveFile
+{
+  ScopedRemoveFile() = default;
+  explicit ScopedRemoveFile(string const & fileName) : m_file(fileName) {}
+
+  ~ScopedRemoveFile()
+  {
+    if (!m_file.empty())
+      std::remove(m_file.c_str());
+  }
+
+  std::string m_file;
+};
+
+static string ReadFileAsString(string const & filePath)
+{
+  ifstream ifs(filePath, ifstream::in);
+  if (!ifs.is_open())
+    return {};
+
+  return {istreambuf_iterator<char>(ifs), istreambuf_iterator<char>()};
+}
+
+
+string RunCurl(string const & cmd)
+{
+  FILE * pipe = ::popen(cmd.c_str(), "r");
+  ASSERT(pipe, ());
+  array<char, 8 * 1024> arr;
+  string result;
+  size_t read;
+  do
+  {
+    read = ::fread(arr.data(), 1, arr.size(), pipe);
+    if (read > 0)
+    {
+      result.append(arr.data(), read);
+    }
+  } while (read == arr.size());
+
+  auto const err = ::pclose(pipe);
+  // Exception will be cought in RunHTTPRequest
+  if (err)
+    throw PipeCallError("", "Error " + strings::to_string(err) + " while calling " + cmd);
+
+  return result;
+}
+
+string GetTmpFileName()
+{  
+  boost::uuids::random_generator gen;
+  boost::uuids::uuid u = gen();
+
+  stringstream ss;
+  ss << u;
+
+  ASSERT(!ss.str().empty(), ());
+
+  return GetPlatform().TmpPathForFile(ss.str());
+}
+
+typedef vector<pair<string, string>> HeadersT;
+
+HeadersT ParseHeaders(string const & raw)
+{
+  istringstream stream(raw);
+  HeadersT headers;
+  string line;
+  while (getline(stream, line))
+  {
+    auto const cr = line.rfind('\r');
+    if (cr != string::npos)
+      line.erase(cr);
+
+    auto const delims = line.find(": ");
+    if (delims != string::npos)
+      headers.push_back(make_pair(line.substr(0, delims), line.substr(delims + 2)));
+  }
+  return headers;
+}
+}  // namespace
+// Used as a test stub for basic HTTP client implementation.
+// Make sure that you have curl installed in the PATH.
+// TODO(AlexZ): Not a production-ready implementation.
+namespace platform
+{
+// Extract HTTP headers via temporary file with -D switch.
+// HTTP status code is extracted from curl output (-w switches).
+// Redirects are handled recursively. TODO(AlexZ): avoid infinite redirects loop.
+bool HttpClient::RunHttpRequest()
+{
+  ScopedRemoveFile headers_deleter(GetTmpFileName());
+  ScopedRemoveFile body_deleter;
+  ScopedRemoveFile received_file_deleter;
+
+  string cmd = "curl -s -w '%{http_code}' -X " + m_httpMethod + " -D '" + headers_deleter.m_file + "' ";
```

Comment:
I still strongly suggest you to investigate how Qt Http request works. And if it can be used here, for synchronous http request, please, do that instead. But this is a non-blocking comment, so, feel free to leave this code as is.